### PR TITLE
Try async base container

### LINKED ISSUE #1253: Get rid of async code in BaseContaner's constructor
We run async code in BaseContaner's constructor which isn't good, and actually not possible with async/await. We need to figure out how to re-write it in async way. This error is thrown:

```
(node:15010) UnhandledPromiseRejectionWarning: TimeoutError: Timed out waiting for element with css selector of '.plans-features-main__group' to be present and displayed
Wait timed out after 40821ms
    at /Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:2201:17
    at ManagedPromise.invokeCallback_ (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:1376:14)
    at TaskQueue.execute_ (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:3084:14)
    at TaskQueue.executeNext_ (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:3067:27)
    at asyncRun (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:2927:27)
    at /Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:668:7
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
From: Task: Timed out waiting for element with css selector of '.plans-features-main__group' to be present and displayed
    at scheduleWait (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:2188:20)
    at ControlFlow.wait (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:2517:12)
    at Driver.wait (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/webdriver.js:934:29)
    at Object.waitTillPresentAndDisplayed (/Developer/a8c/wp-e2e-tests/lib/driver-helper.js:106:16)
    at PickAPlanPage.waitForPage (/Developer/a8c/wp-e2e-tests/lib/base-container.js:50:29)
    at new BaseContainer (/Developer/a8c/wp-e2e-tests/lib/base-container.js:28:8)
    at new PickAPlanPage (/Developer/a8c/wp-e2e-tests/lib/pages/signup/pick-a-plan-page.js:13:3)
    at Context.<anonymous> (/Developer/a8c/wp-e2e-tests/specs-jetpack-calypso/wp-jetpack-connect-spec.js:145:25)
    at runTest (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/testing/index.js:163:22)
    at /Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/testing/index.js:184:16
    at new ManagedPromise (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:1077:7)
    at controlFlowExecute (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/testing/index.js:183:14)
    at TaskQueue.execute_ (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:3084:14)
    at TaskQueue.executeNext_ (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:3067:27)
    at asyncRun (/Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:2974:25)
    at /Developer/a8c/wp-e2e-tests/node_modules/selenium-webdriver/lib/promise.js:668:7
    at <anonymous>
(node:15010) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15010) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### PR DESCRIPTION
**Pull request by alisterscott**
fixes #1253

Adds a `AsyncBaseContainer` which allows us to progressively change pages to support async initialisation using static async methods.

I have updated two pages to use this as an example to begin with.

### CODE REVIEW

**Comment by alisterscott** [comment]
Looks like this is passing now

**Comment by brbrr** [comment]
I like the changes! 

What do you think about moving these static init methods into `AsyncBaseContainer`? You can use `this` instead of class name, which will make that method to work will all inherited classes. Here is an example code explaining this concept:

```
class A {
	constructor() {
		this.name = 'class ' + this.constructor.name;
	}

	static init() {
		const klass = new this();
		return klass;
	}
}

class B extends A {
}

class C extends B {
}

A.init(); // => A { name: 'class A' }
B.init(); // => B { name: 'class B' }
C.init(); // => C { name: 'class C' }
```

**Comment by alisterscott** [comment]
> Is there a possibility to change existing BaseContainer so that all pages can use it immediately?

I don't know how to do this without changing every single page all at once which I don't want to do - the PR will be too big and too much risk of getting something slightly wrong. Do you know of a way?

**Comment by alisterscott** [comment]
@brbrr Thanks for the feedback - can you take another look? It all lives in the base container now so there should be less code as we apply this to other pages 👍

File: lib/async-base-container.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		await this.waitForPage();
+		await this.checkForUnknownABTestKeys();
+		return await this.checkForConsoleErrors();
+	}
+
+	async waitForPage() {
+		return await driverHelper.waitTillPresentAndDisplayed(
+			this.driver,
+			this.expectedElementSelector,
+			this.explicitWaitMS
```
Comment (brbrr) [inline lib/async-base-container.js]: passing default wait timeout is not needed here

File: lib/async-base-container.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		);
+	}
+
+	async displayed() {
+		return await driverHelper.isEventuallyPresentAndDisplayed(
+			this.driver,
+			this.expectedElementSelector,
+			this.explicitWaitMS
```
Comment (brbrr) [inline lib/async-base-container.js]: Same as above

File: lib/async-base-container.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import * as slackNotifier from './slack-notifier';
+
+export default class AsyncBaseContainer {
+	constructor( driver, expectedElementSelector, url = null ) {
+		this.name = this.constructor.name;
+		this.driver = driver;
+		this.screenSize = driverManager.currentScreenSize().toUpperCase();
+		this.expectedElementSelector = expectedElementSelector;
+		this.url = url;
+	}
```
Comment (brbrr) [inline lib/async-base-container.js]: I don't see that `this.explicitWaitMS` is defined somewhere. But it is used below and in other pages. Should it be added into the constructor?

File: specs/wp-signup-spec.js [hunk 4]
```diff
@@ -81,7 +81,7 @@ test.describe( `[${ host }] Sign Up  (${ screenSize }, ${ locale })`, function()
 				return await driverManager.ensureNotLoggedIn( driver );
 			} );
 
-			test.it( 'Can visit the start page', async function() {
+			test.it( 'driver the start page', async function() {
```
Comment (brbrr) [inline specs/wp-signup-spec.js]: was it done intentionally?

File: specs/wp-signup-spec.js [hunk 5]
```diff
@@ -192,10 +193,12 @@ test.describe( `[${ host }] Sign Up  (${ screenSize }, ${ locale })`, function()
 					const closeAccountPage = await new CloseAccountPage( driver );
 					await closeAccountPage.chooseCloseAccount();
 					await closeAccountPage.enterAccountNameAndClose( blogName );
-					return await new LoggedOutMasterbarComponent( driver ).displayed();
+					const loggedOutMasterbarComponent = await LoggedOutMasterbarComponent.Expect( driver );
+					return await loggedOutMasterbarComponent.displayed();
```
Comment (brbrr) [inline specs/wp-signup-spec.js]: all these `.displayed()` calls looks like not needed

**Comment by alisterscott** [comment]
I just realized the wp-admin login page is now broken since it extends the logon page which I've changed - I need to address this now 🤔

**Comment by alisterscott** [comment]
@brbrr  thanks for your feedback so far - do you mind taking another look with my changes to the wp-admin login page (which were necessary since this used to extend the calypso login page which changed as part of this PR)

@bsessions85 do you mind taking a look also, I want to make sure you're ok with the `Visit` and `Expect` methods and this slightly new approach

File: lib/pages/wp-admin/wp-admin-logon-page.js [hunk 6]
```diff
@@ -3,42 +3,19 @@
...
 		}
-
-		super( driver, visit, wpAdminURL );
-		if ( forceStandardLogon === true ) {
-			this.ensureWPAdminStandardLogonShown();
-		}
+		super( driver, by.css( '.login' ), wpAdminURL );
 	}
 
-	ensureWPAdminStandardLogonShown() {
```
Comment (alisterscott) [inline lib/pages/wp-admin/wp-admin-logon-page.js]: this wasn't used

File: lib/pages/wp-admin/wp-admin-logon-page.js [hunk 7]
```diff
@@ -3,42 +3,19 @@
...
+export default class WPAdminLogonPage extends AsyncBaseContainer {
+	constructor( driver, url ) {
+		let wpAdminURL = null;
 		if ( url ) {
 			url = url.replace( /^https?:\/\//, '' ).replace( /\/wp-admin/, '' );
 			wpAdminURL = `http://${ url }/wp-admin`;
 		}
-
-		super( driver, visit, wpAdminURL );
-		if ( forceStandardLogon === true ) {
```
Comment (alisterscott) [inline lib/pages/wp-admin/wp-admin-logon-page.js]: this wasn't used

**Comment by alisterscott** [comment]
I'm going to merge this one whilst it's still fresh - I tested a few different Jetpack scenarios locally but please keep any eye out for any issues @brbrr 

@Automattic/flowpatrol if you have any feedback on this I can take it on board and create a new PR before we go about converting all the pages over in different PRs

**Comment by bsessions85** [comment]
This looks good to me. good work!

**Comment by Stojdza** [comment]
> the PR will be too big and too much risk of getting something slightly wrong

I agree. Let's go step by step. Anyway, awesome work!

@brbrr I put additional `.displayed()` calls in several places when I disabled Promise manager. That was the way to properly load some pages/components. I hope that we don't need them anymore with this new AsyncBaseContainer.

**Pull request merged**